xmake/rules/c++/modules/xmake.lua

```diff
@@ -117,43 +110,70 @@ rule("c++.build.modules.builder")
     -- serial compilation only, usually used to support project generator
     before_buildcmd_files(function(target, batchcmds, sourcebatch, opt)
         if target:data("cxx.has_modules") then
-            import("modules_support.common")
+            import("modules_support.compiler_support")
+            import("modules_support.dependency_scanner")
+            import("modules_support.builder")
+
+            -- add target deps modules
+            if target:orderdeps() then
+                local deps_sourcefiles = dependency_scanner.get_targetdeps_modules(target)
+                if deps_sourcefiles then
+                    table.join2(sourcebatch.sourcefiles, deps_sourcefiles)
+                end
+            end
+
+            -- append std module
+            table.join2(sourcebatch.sourcefiles, compiler_support.get_stdmodules(target) or {})
+
+            -- extract packages modules dependencies
+            local package_modules_data = dependency_scanner.get_all_packages_modules(target, opt)
+            if package_modules_data then
+                -- append to sourcebatch
+                for _, package_module_data in pairs(package_modules_data) do
+                    table.insert(sourcebatch.sourcefiles, package_module_data.file)
+                    target:fileconfig_add(package_module_data.file, {external = true, defines = package_module_data.metadata.defines})
+                end
+            end
 
-            -- patch sourcebatch
-            common.patch_sourcebatch(target, sourcebatch, opt)
+            compiler_support.patch_sourcebatch(target, sourcebatch, opt)
+            local modules = dependency_scanner.get_module_dependencies(target, sourcebatch, opt)
 
-            -- generate headerunits
-            local modules = common.get_module_dependencies(target, sourcebatch, opt)
-            common.generate_headerunits_for_batchcmds(target, batchcmds, sourcebatch, modules, opt)
+            opt.batchjobs = false
+
+            -- build headerunits
+            builder.build_headerunits_for_batchcmds(target, batchcmds, sourcebatch, modules, opt)
 
             -- build modules
-            common.build_modules_for_batchcmds(target, batchcmds, sourcebatch, modules, opt)
+            builder.build_modules_for_batchcmds(target, batchcmds, sourcebatch, modules, opt)
+
+            -- cull external modules objectfile
+            compiler_support.cull_objectfiles(target, sourcebatch)
         else
             -- avoid duplicate linking of object files of non-module programs
             sourcebatch.objectfiles = {}
         end
     end)
 
     before_link(function (target)
-        import("modules_support.common")
+        import("modules_support.builder")
         if target:data("cxx.has_modules") then
-            common.append_dependency_objectfiles(target)
+            -- builder.append_dependency_objectfiles(target)
```

Comment:
?

Reply:
thx, i forgot to remove the whole before_link